python/src/iceberg/avro/file.py

```diff
@@ -0,0 +1,160 @@
+        field_id=200,
+        name="meta",
+        field_type=MapType(key_id=201, key_type=StringType(), value_id=202, value_type=BinaryType(), value_is_optional=False),
+        is_optional=False,
+    ),
+    NestedField(field_id=300, name="sync", field_type=FixedType(length=SYNC_SIZE), is_optional=False),
+)
+
+
+@dataclass(frozen=True)
+class AvroFileHeader:
+    magic: bytes
+    meta: dict[str, str]
+    sync: bytes
+
+    def get_compression_codec(self) -> type[Codec]:
+        """Get the file's compression codec algorithm from the file's metadata."""
+        codec_key = "avro.codec"
+        if codec_key in self.meta:
+            codec_name = self.meta[codec_key]
+
+            if codec_name not in KNOWN_CODECS:
+                raise ValueError(f"Unsupported codec: {codec_name}. (Is it installed?)")
+
+            return KNOWN_CODECS[codec_name]
+        else:
+            return NullCodec
+
+    def get_schema(self) -> Schema:
+        schema_key = "avro.schema"
+        if schema_key in self.meta:
+            avro_schema_string = self.meta[schema_key]
+            avro_schema = json.loads(avro_schema_string)
+            return AvroSchemaConversion().avro_to_iceberg(avro_schema)
+        else:
+            raise ValueError("No schema found in Avro file headers")
+
+
+class AvroFile:
+    input_file: InputFile
+    input_stream: InputStream
+    header: AvroFileHeader
+    schema: Schema
+    file_length: int
+
+    decoder: BinaryDecoder
+    block_decoder: BinaryDecoder
+
+    block_records: int = -1
+    block_position: int = 0
+    block: int = 0
+
+    def __init__(self, input_file: InputFile) -> None:
+        self.input_file = input_file
+
+    def __enter__(self):
+        """
+        Opens the file and reads the header and generates
+        a reader tree to start reading the payload
+
+        Returns:
+            A generator returning the AvroStructs
+        """
+        self.input_stream = self.input_file.open()
+        self.decoder = BinaryDecoder(self.input_stream)
+        self.header = self._read_header()
+        self.schema = self.header.get_schema()
+        self.file_length = len(self.input_file)
+        self.reader = visit(self.schema, ConstructReader())
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.input_stream.close()
+
+    def __iter__(self) -> AvroFile:
+        return self
+
+    def _read_block(self) -> int:
+        if self.block > 0:
+            sync_marker_len = len(self.header.sync)
+            sync_marker = self.decoder.read(sync_marker_len)
+            if sync_marker != self.header.sync:
+                raise ValueError(f"Expected sync bytes {self.header.sync!r}, but got {sync_marker!r}")
+            if self.is_EOF():
+                raise StopIteration
+        self.block = self.block + 1
+        self.block_records = self.decoder.read_long()
+        self.block_decoder = self.header.get_compression_codec().decompress(self.decoder)
```

Comment:
It's a bit odd to me that the decompression codec accepts and produces a decoder, rather than decompressing bytes and returning bytes. It's okay for now, but I think it mixes Avro-specific classes into the compression classes, which don't need to be specific to Avro.